### PR TITLE
Example that describes lambdas and std::function as kernel functions

### PR DESCRIPTION
**Pull request by erikzenker**
I did not know that lambdas and std::function can be used within alpaka
until I read your thesis more precisely. So I tried it and it worked !
Finally I wrote this example which is a small modification of the
Hello World example.

### CODE REVIEW

**Comment by BenjaminW3** [comment]
Cool that this is really working ;-)

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+                                                          for(size_t i = 0; i < nExclamationMarks; ++i){
+                                                              printf("!");
+                                                          }
+                                                          
+                                                          printf("\n");
+
+                                                      },
+                                                      nExclamationMarks // 1st real kernel argument, but 2nd argument in the lambda function,
+                                                                        // since, the first argument as allways the accelerator!
```
Comment (BenjaminW3) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: ~~as~~ -> is

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                                                      ));
+
+    alpaka::stream::enqueue(stream, helloWorld);
+
+    
+    /**
+     * Run kernel with std::function
+     *
+     * This kernel says hi to world by using
```
Comment (BenjaminW3) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: to **the** world

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+     * Alpaka accepts these std::functions 
+     * as kernel functions. Therefore, it is easy
+     * to wrap allready existing code into a
+     * std::function and provide it to the alpaka 
+     * library.
+     */
+    auto const hiWorld (alpaka::exec::create<Acc> (workdiv,
+                                                   std::function<void(Acc&, size_t)>( hiWorldFunction<Acc> ),
+                                                   nExclamationMarks)); // 1st real kernel argument, but 2nd argument in the wrapped function
+                                                                        // since, the first argument as allways the accelerator!
```
Comment (BenjaminW3) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: ~~as~~ -> is

**Comment by BenjaminW3** [comment]
When you have corrected the typos and squashed the commits this is ready to merge.

**Comment by BenjaminW3** [comment]
Oh no, wait. The CI tells me that CUDA 7.0 does not correctly support C++11 lambda functions even in Host code...
Furthermore, the test is only using the `DevCpu` which means that it does not show that it is also woking in CUDA. I had already written a possible solution in #47 where `ALPAKA_FN_ACC` has to be added after the lambda capture list as well as a specific nvcc compiler option. Because the host only definition of `ALPAKA_FN_ACC` is currently not empty, this could possibly require further changes.

**Comment by erikzenker** [comment]
Mmh I am not able to compile when the cuda acc is enabled at all, also not the other examples

**Comment by erikzenker** [comment]
"Falscher alarm" just had a fix on the cluster so piconalpaka can compile and run

    -    #define ALPAKA_FN_ACC __device__ __host__ __forceinline__
    +    #define ALPAKA_FN_ACC __device__ __forceinline__

**Comment by erikzenker** [comment]
I used the following lambda : 

    auto foo = [] __device__ (Acc& acc) {
        auto globalThreadIdx    = alpaka::idx::getIdx<alpaka::Grid, alpaka::Threads>(acc);
    };
    
This lambda function is just written in the code and not used, but produces the following error
during compilation. So we also have a problem with host and device functions of cuda.

`error: calling a __device__ function("getIdx") from a __host__ __device__ function("getIdx") is not allowed`

**Comment by erikzenker** [comment]
Why does this 
   global -> host device -> device

call does work for non lambda kernels ?

**Comment by erikzenker** [comment]
My last comment is related to #78  ...lambdas are also functions without further templates

**Comment by erikzenker** [comment]
This PR is also related to #73.

File: example/helloWorldLambda/src/helloWorldLambda.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+     * lambda function. This kind of kernel function
+     * declaration might be useful when small kernels
+     * are written for testing or lambda functions
+     * allready exist.
+     *
+     */
+    const size_t nExclamationMarks = 10;
+    
+    auto const helloWorld (alpaka::exec::create<Acc> (workdiv,
+                                                      [](Acc& acc, size_t const nExclamationMarks) -> void {
```
Comment (psychocoderHPC) [inline example/helloWorldLambda/src/helloWorldLambda.cpp]: Please use indention instead of alignment

**Comment by erikzenker** [comment]
After a further discussion with @psychocoderHPC we agree that it is not possible to use lambdas in alpaka in combination with the cuda accelerator. It might be possible if we change the ALPAKA_FN... defintion
but this leads to the problem that we can not use a cpu acc at the same time. I can live without lambdas, but it would be a nice to have.

**Comment by BenjaminW3** [comment]
We could maybe use C++14 generic lambdas with auto as parameter type.

```C++
auto l = [] (auto const &  TAcc) { ...; };
```

So that l is an instance of the type:

```C++
class /* unnamed */
{
public:
    template<
        typename TAcc>
    auto operator() (auto const & TAcc) const
    -> void
    { ...; }
};
```

**Comment by erikzenker** [comment]
In world were nvcc accepts C++14 lambdas this could be a solution (but we do not know if this does the trick). From my point of view we can close this PR.

**Pull request closed without merging**